docs/weave-tutorial/06-buckets.md

```diff
@@ -0,0 +1,131 @@
+        morm.WithIndex("marketWithTickers", marketIDTickersIndexer, true),
+    )
+
+    return &OrderBookBucket{
+        ModelBucket: b,
+    }
+}
+```
+
+Let's explain what the heck is `morm.WithIndex("market", marketIDindexer, false)`?
+
+## Secondary Indexes
+
+Sometimes we need another index for the data. Generally, we will look up an order book from the market it belongs to and it’s index in the market. But what if we want to list all orderbooks of a market overall order books? For this, we need to add a secondary index on the order books to query by market. This is a typical case and weave provides nice support for this functionality.
+
+We add an indexing method to take any object, enforce the type to be a proper Orderbook, then extract the index we want. This can be a field or any deterministic transformation of one (or multiple) fields. The output of the index becomes key in another query. Bucket provides a simple method to query by index.
+
+**marketIDindexer** is a secondary index with only market ID. This is a simple index form to implement.
+It binds OrderBook to a MarketId(*bytes*)
+
+Weave uses uniformed *bytes* as indexes. This improves performance.
+
+```go
+func marketIDindexer(obj orm.Object) ([]byte, error) {
+    if obj == nil || obj.Value() == nil {
+        return nil, nil
+    }
+    ob, ok := obj.Value().(*OrderBook)
+    if !ok {
+        return nil, errors.Wrapf(errors.ErrState, "expected orderbook, got %T", obj.Value())
+    }
+    return ob.MarketID, nil
+}
+```
+
+You must have ideas flying around in your mind like **how are we going to make a compound index? Really!? Is it all weave has?**
+
+Don't worry. Weave is like a swiss knife with a lot of blockchain features.
+
+Here is how we create compound index for morm buckets:
+
+```go
+func BuildMarketIDTickersIndex(orderbook *OrderBook) []byte {
+    askTickerByte := make([]byte, tickerByteSize)
+    copy(askTickerByte, orderbook.AskTicker)
+
+    bidTickerByte := make([]byte, tickerByteSize)
+    copy(bidTickerByte, orderbook.BidTicker)
+
+    return bytes.Join([][]byte{orderbook.MarketID, askTickerByte, bidTickerByte}, nil)
+}
+```
+
+*BuildMarketIDTickersIndex = indexByteSize = 8(MarketID) + ask ticker size + bid ticker size*
+
+Sample market id index with tickers = `000000056665820070797900`
+
+`000000056665820070797900` = 00000005(**MarketID = 5**) + 6665200(**BAR ticker in bytes**) + 70797900(**FOO ticker in bytes**)
+
+## Custom Buckets
+
+Data consistency must be enforced on the buckets. All data is validated before saving, but we also need to make sure that all data is the proper type of object before saving. Unfortunately, this is quite difficult to do compile-time without generic, so a typical approach is to embed the orm.Bucket in another struct and just force validation of the object type runtime before save. Wrap an orm.Bucket with the functionalities you desire, there you have a custom bucket to serve your needs.
```

Comment:
> be enforced on ~the~ buckets

Reply:
So you say custom buckets are kind of unnecessary or does it mean custom need buckets must be implemented `orm.ModelBucket`  not `orm.Bucket`? Do I understand correct?

Reply:
Yes. When using `orm.ModelBucket` there is no need to extend it. It provides a complete set of functionalities.

As an example, look at [`VoteBucket.GetVote`](https://github.com/iov-one/weave/blob/master/x/gov/bucket.go#L245) - it is a wrapper over `Bucket.Get` that does type checking and casting. This code is not needed when using `orm.ModelBucket` beucase the way you [use it implies the right type](https://github.com/iov-one/weave/blob/master/x/distribution/handler.go#L106-L109).